mw_validate_json.go

```diff
@@ -0,0 +1,81 @@
+package main
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/xeipuuv/gojsonschema"
+
+	"github.com/TykTechnologies/tyk/apidef"
+)
+
+type ValidateJSON struct {
+	BaseMiddleware
+}
+
+func (k *ValidateJSON) Name() string {
+	return "ValidateJSON"
+}
+
+func (k *ValidateJSON) EnabledForSpec() bool {
+	for _, v := range k.Spec.VersionData.Versions {
+		if len(v.ExtendedPaths.ValidateJSON) > 0 {
+			return true
+		}
+	}
+
+	return false
+}
+
+// ProcessRequest will run any checks on the request on the way through the system, return an error to have the chain fail
+func (k *ValidateJSON) ProcessRequest(w http.ResponseWriter, r *http.Request, _ interface{}) (error, int) {
+
+	_, versionPaths, _, _ := k.Spec.Version(r)
+	found, meta := k.Spec.CheckSpecMatchesStatus(r, versionPaths, ValidateJSONRequest)
+	if !found {
+		return nil, 200
+	}
+
+	mmeta := meta.(*apidef.ValidatePathMeta)
+	if mmeta.ValidateWith64 == "" {
+		return errors.New("no schemas to validate against"), 400
+	}
+
+	rCopy := copyRequest(r)
+	bodyBytes, err := ioutil.ReadAll(rCopy.Body)
+	if err != nil {
+		return err, 400
+	}
+	defer rCopy.Body.Close()
+
+	schemaBytes, err := base64.StdEncoding.DecodeString(mmeta.ValidateWith64)
+	if err != nil {
+		return errors.New("unable to base64 decode schema"), 400
+	}
+
+	result, err := k.validate(bodyBytes, schemaBytes)
+	if err != nil {
+		return err, 400
```

Comment:
might worth to log err so we will be able to explain to customer why it was `400`

Reply:
@dencoded - Agree with status code changes and have implemented these.
The reason I have not logged before returning error is that IMO it's bad practice to log the error before returning. It's up to the caller of `ProcessRequest` to do the logging if it needs to. Open to thoughts on this.